lib/broadway_kafka/producer.ex

```diff
@@ -524,30 +529,70 @@ defmodule BroadwayKafka.Producer do
   end
 
   defp maybe_schedule_poll(state, interval) do
-    %{buffer: buffer, demand: demand, acks: acks, receive_timer: receive_timer} = state
+    %{
+      buffers: buffers,
+      demand: demand,
+      max_demand: max_demand,
+      acks: acks,
+      receive_timer: receive_timer
+    } = state
 
-    case dequeue_many(buffer, acks, demand, []) do
-      {acks, 0, events, buffer} ->
-        {:noreply, events, %{state | demand: 0, buffer: buffer, acks: acks}}
+    {key, next_queue, next_index} = find_next_queue(buffers, state.buffer_index)
 
-      {acks, demand, events, buffer} ->
-        receive_timer = receive_timer || schedule_poll(state, interval)
+    num_to_dequeue = min(demand, max_demand)
+    {updated_queue, acks, events} = dequeue_many(next_queue, acks, key, num_to_dequeue, [])
+    buffers = put_queue(buffers, key, updated_queue)
+    new_demand = demand - length(events)
 
-        state = %{
-          state
-          | demand: demand,
-            buffer: buffer,
-            receive_timer: receive_timer,
-            acks: acks
-        }
+    receive_timer =
+      if new_demand > 0 do
+        receive_timer || schedule_poll(state, interval)
+      else
+        receive_timer
+      end
+
+    state = %{
+      state
+      | receive_timer: receive_timer,
+        buffer_index: next_index,
+        acks: acks,
+        buffers: buffers,
+        demand: new_demand
+    }
 
-        {:noreply, events, state}
+    {:noreply, events, state}
+  end
+
+  defp find_next_queue(queues, buffer_index) do
```

Comment:
TODO: Use ring approch suggested in the issue

Also I found if some queue are empty, we are using the cycle without picking up any event, this function tries to skip the empty queues or fallback to the first item

Reply:
I'm thinking in to add an optional option `max_pending_acks` in a separate PR. With that we can skip the queues that have more than requested pending acks. 

It's necessary to avoid endless rebalancing issue in my case.